packages-exp/auth-exp/src/core/user/additional_user_info.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ * @license
+ * Copyright 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  AdditionalUserInfo,
+  UserProfile,
+  ProviderId
+} from '@firebase/auth-types-exp';
+import { IdTokenResponse, IdTokenResponseKind } from '../../model/id_token';
+import { _parseToken } from './id_token_result';
+import { assert } from '../util/assert';
+
+/**
+ * Parse the `AdditionalUserInfo` from the ID token response.
+ */
+export function fromIdTokenResponse(
+  idTokenResponse: IdTokenResponse,
+  appName: string
+): AdditionalUserInfo | null {
+  const { providerId } = idTokenResponse;
+  const profile = idTokenResponse.rawUserInfo
+    ? JSON.parse(idTokenResponse.rawUserInfo)
+    : {};
+  const isNewUser =
+    !!idTokenResponse.isNewUser ||
+    idTokenResponse.kind === IdTokenResponseKind.SignupNewUser;
+  if (!providerId && !!idTokenResponse) {
+    const providerId = _parseToken(idTokenResponse.idToken)?.firebase?.[
+      'sign_in_provider'
+    ];
+    assert(
+      // @ts-ignore - Check to see if string is castable to enum.
+      !providerId || Object.values(ProviderId).includes(providerId),
```

Comment:
this won't work anymore since we're using const enums, you can probably remove this assertion now until we figure out a better way, also this check doesn't work with federated providers